libretro core

warmenhoven opened the pull request:
For context I didn't actually write most of this, it was brought over from the citra libretro core. My goal is to support it and improve on it over time.

I've tried to trim down the changes to as small as I thought I could get away with, in the hopes that it could be merged into this repo, and make it easier to maintain long term.

There are some changes, particularly to the CMakeLists.txt files, that are there to satisfy the libretro buildbots. I'm happy to get into the details of exactly what the buildbots are, but the basic point is that they frequently use outdated tools and toolchains in order to support older OS versions or more esoteric hardware. You'll also note that there is the addition of a .gitlab-ci.yml file, for the purpose of integrating with the libretro buildbot. I've currently already got it building there, https://git.libretro.com/warmenhoven/azahar/-/pipelines/369383.

~~There are still a number of major bugs that need to be addressed before this can reasonably be merged, but I wanted to submit a draft PR early to start getting feedback.~~

- [x] ~~GitHub action to build core~~
- [x] ~~context_reset for gl is not correct - when toggling fullscreen, android backgrounding, etc., causes black screen/distorted screen/audio lags/audio distortions~~
- [x] ~~vulkan renderer crashes when paused - this only seems to happen on some builds? I have seen builds where this works fine, I don't know why~~
- [x] ~~vulkan sync issues? seems to be causing visual artifacts and hangs, esp. on android~~
- [x] ~~tvOS crashes on load - seems to initialize vulkan context fine?~~
- [x] ~~core options need to be moved to options_v2~~
- [ ] core options are incomplete
- [ ] screen layout options are incomplete
- [x] ~~iOS/tvOS need to force cpu_jit off if unavailable~~
- [x] ~~retro_get_memory_* is incomplete~~
- [x] ~~touchscreen on second screen is not hooked up (mouse/touch input from frontend)~~
- [x] ~~gyro is not hooked up~~
- [ ] right analog as gyro?
- [x] ~~microphone is not hooked up~~
- [ ] camera is not hooked up

PabloMK7 commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,141 @@
+    JNI_PATH: .
+    CORENAME: azahar
+    API_LEVEL: 21
+    BASE_CORE_ARGS: -DENABLE_LIBRETRO=ON -DENABLE_SDL2=OFF -DENABLE_QT=OFF -DENABLE_TESTS=OFF -DENABLE_ROOM=OFF -DENABLE_WEB_SERVICE=OFF -DENABLE_SCRIPTING=OFF -DENABLE_CUBEB=OFF -DENABLE_OPENAL=OFF -DENABLE_LIBUSB=OFF -DCITRA_WARNINGS_AS_ERRORS=OFF
```
Why disable scripting? It can be controlled by config, and it should be usable.

warmenhoven commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,141 @@
+    JNI_PATH: .
+    CORENAME: azahar
+    API_LEVEL: 21
+    BASE_CORE_ARGS: -DENABLE_LIBRETRO=ON -DENABLE_SDL2=OFF -DENABLE_QT=OFF -DENABLE_TESTS=OFF -DENABLE_ROOM=OFF -DENABLE_WEB_SERVICE=OFF -DENABLE_SCRIPTING=OFF -DENABLE_CUBEB=OFF -DENABLE_OPENAL=OFF -DENABLE_LIBUSB=OFF -DCITRA_WARNINGS_AS_ERRORS=OFF
```
Frontends are already able to implement their own similar scripting functionality (RetroArch has already done this). It shouldn't be necessary to handle it at the core layer.

warmenhoven replied:
Hm, actually, it might be possible to do something more correct using `RETRO_ENVIRONMENT_SET_MEMORY_MAPS`. Let me look into that a bit more.

warmenhoven commented on CMakeLists.txt:
```diff
@@ -1,6 +1,10 @@
+if(ANDROID)
+  cmake_minimum_required(VERSION 3.25)
+else()
+  cmake_minimum_required(VERSION 3.23)
```
Second question first, this is the Dockerfile that was used to create the android build container: https://git.libretro.com/libretro-infrastructure/libretro-build-android/-/blob/master/Dockerfile?ref_type=heads. You can pull down the docker image from git.libretro.com:5050/libretro-infrastructure/libretro-build-android:latest

I've verified that the core runs :) Yes I do believe that LTO is correctly enabled for Android.

PabloMK7 replied:
That's actually the reason we bumped the cmake version. It's not that gold is missing, it's a misdetect in older cmake versions. Newer android build environments use a different linker and older cmake wasn't detecting them.

warmenhoven replied:
Hm, it is using cmake 3.25, and that's the most recent version it can easily be updated to. Let me do a little more digging here.

PabloMK7 replied:
The gold linker is deprecated, that's why it's no longer there. CMake should accomodate to it, but there was a bug in versions prior to 3.25 that still tried to use gold instead of the new lld.

See https://github.com/azahar-emu/azahar/pull/1151

warmenhoven commented on CMakeLists.txt:
```diff
@@ -24,13 +28,15 @@ else()
 endif()
 
-if (APPLE)
+if (APPLE AND NOT LIBRETRO)
```
🤦 Yes, thank you.

warmenhoven commented on src/CMakeLists.txt:
```diff
@@ -110,10 +110,14 @@ else()
+    if (NOT ENABLE_LIBRETRO)
+        add_compile_options(
+            -Wp,-D_GLIBCXX_ASSERTIONS
+            -fstack-clash-protection
```
The compilers used in the libretro buildbot don't support those compiler options. I can provide more details about specifically which compilers were failing with this if desired.

warmenhoven commented on src/citra_libretro/citra_libretro.cpp:
```diff
@@ -0,0 +1,791 @@
+    LOG_DEBUG(Frontend, "Initialising core...");
+
+    // Set up LLE cores
+    for (const auto& service_module : Service::service_module_map) {
```
I will research this and the following comments on citra_libretro.cpp and get back to you. As I mentioned, I just brought over the code from citra and got it running; I haven't gone through and done a thorough audit of it myself yet. That will definitely happen before I take this PR out of draft state.

warmenhoven commented on src/citra_libretro/citra_libretro.cpp:
```diff
@@ -0,0 +1,791 @@
+
+    Settings::values.current_input_profile.touch_device = "engine:emu_window";
+
+    // Hardcode buttons to bind to libretro - it is entirely redundant to have
```
Sure, though, I'm probably going to end up changing this up slightly before this gets merged.

The frontend handles all input and controller drivers and remapping. The core shouldn't think about any of it.

Instead it just needs to tell the frontend what types of ports the original hardware had and what types of inputs it supported on each of those ports (https://github.com/warmenhoven/azahar/blob/0141745abdcd1d2ec14a2db7aa481d60fd1b254b/src/citra_libretro/citra_libretro.cpp#L177). Then for each type of input, it needs to provide an input descriptor along with how each is mapped from the libretro RetroPad API (https://github.com/warmenhoven/azahar/blob/0141745abdcd1d2ec14a2db7aa481d60fd1b254b/src/citra_libretro/citra_libretro.cpp#L220).

I'm happy to go into more detail on this if you'd like, I feel like this description really only scratches the surface of a lot of underlying details.

Algiuxs commented:
Testing on android RetroArch nightly, touch doesn't work, I can interact with RetroArch controls, but can't click elements that are on screen

Algiuxs commented:
Also can't get to RetroArch menu after launching any 3ds game on Azahar core on Mali device, the app just crashes.

warmenhoven commented:
Thanks very much for testing it out! Yeah, I know about those. I did say there are a number of major bugs. :)

Sorry, I should probably put my list of known issues and todo items someplace public. I think what I'll do is go back and edit the original comment on the PR and put in the list there.

KojoZero commented:
I've been trying to build this using the MSVC Instructions on Windows 11 with ENABLE_LIBRETRO on, but I've been getting these errors.

[azahar_buildlog.txt](https://github.com/user-attachments/files/21478773/azahar_buildlog.txt)

warmenhoven commented:
@KojoZero Neat, I'm impressed you got that far with it. I was never able to get the MSVC building back when it was citra and ended up using mingw for it, and never went back to try to get it working.

See also:
https://github.com/warmenhoven/azahar/blob/bcd1522c5a0eae67394b936bfee331cc4692bb05/.gitlab-ci.yml#L18
https://github.com/warmenhoven/azahar/blob/bcd1522c5a0eae67394b936bfee331cc4692bb05/.github/workflows/libretro.yml#L76

There's probably several places in the CMakeFiles that assume if you're building the windows libretro core that you're using mingw as a result, and the msvc path is probably not going to work. I'm not really planning on working on that since the buildbot is using mingw, but if you get it working I'm happy to include it in this PR.

KojoZero commented:
If you have time, do you think you could add gyro as another Right Analog Function (Yaw and Pitch controlled by analog stick, and maybe roll controlled by L2/R2)?

warmenhoven commented:
Actually sorry it gets weirder than that, because each axis is +/- but the triggers are only +.

Also the accelerometer includes gravity.

I think there was a way that the citra core could use the mouse for gyro though. Maybe I can look at that.

warmenhoven commented:
I've fixed all of the known major bugs. There is still missing functionality relative to the standalone, and I'm going to keep working on the core regardless of when it's merged. But at this point, it has more functionality than the citra core did, isn't missing any functionality that core had, and works for more platforms. So, I've decided to take it out of "Draft" status and mark it as ready for review.

Please, let me know what the blockers for merging are and I'll address them.

Also, this PR's git history has gotten a little messy. Let me know if you'd like me to clean it up at all. I'm happy to rebase/squash/clean up however you recommend. Or I'm happy to just leave it.

For interested testers, I've created the GitHub action to create the libretro core, and you can download it from there: https://github.com/azahar-emu/azahar/actions/workflows/libretro.yml. (It's also available on the libretro gitlab builder: https://git.libretro.com/warmenhoven/azahar/-/artifacts, but it has a pretty aggressive cleanup policy so if the core hasn't been built in a few days it might not be available there.)

PabloMK7 commented:
Thanks a lot, I will start reviewing in 2 weeks (busy currently). Please remind me in 2 weeks!

Algiuxs commented:
Noticed some things: 
1. I suspect that when shaders aren't fully compiled it shows garbage.
2. RetroArch nightly aarch64 crashes when pausing<!-- Error while uploading Screenrecorder-2025-08-30-20-57-39-50.mp4 -->

Algiuxs commented:
Also: https://github.com/user-attachments/assets/16b5219a-a738-4c78-9079-4518c3a8e3e5

f4mrfaux commented:
Is there a to-do or a road map anywhere I can focus my efforts on to help
you? I'm still working on s pen but I'd be happy to assist when I can.

On Sun, Aug 31, 2025, 8:04 AM Algiuxs ***@***.***> wrote:

> *Algiuxs* left a comment (azahar-emu/azahar#1215)
> <https://github.com/azahar-emu/azahar/pull/1215#issuecomment-3240094107>
>
> Also:
> https://github.com/user-attachments/assets/16b5219a-a738-4c78-9079-4518c3a8e3e5
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/azahar-emu/azahar/pull/1215#issuecomment-3240094107>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIDANIEYXBS4GVEX7G2PH3T3QLQFJAVCNFSM6AAAAACA3B2L5KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGA4TIMJQG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

warmenhoven commented:
@Algiuxs hm, weird, I can reproduce something similar but not quite the same on my iPhone, but not on my Mac. I've added it to my list above.

@f4mrfaux The only real todo list is in the original comment for this PR, which you're very welcome to take a shot at (I suspect in particular the microphone one is relatively straightforward?). If you try out the core and see anything I'm missing, I'm always happy to take bug reports or patches/PRs. Any help is greatly appreciated.

warmenhoven commented:
Ok, I'm pretty sure I fixed the vulkan synchronization issues pretty well, so I think it's in good shape now.

Algiuxs commented:
> Hm, I can't reproduce that on my iPhone. Is it pretty consistent?

The error is present on my phone, specs:
OS: HyperOS 2.0.7.0 (android 14)
CPU: MediaTek Dimensity 1080 (MT6877TT)
GPU: Mali-G68 MC4
Phone: Xiaomi Redmi Note 12 Pro 5G

Starman99x commented:
@warmenhoven Can you add support for [compressed 3DS backups](https://azahar-emu.org/blog/compressed-backups/)? I tried to run .z3ds / .z3dsx files but the core does not support them out of the box.

warmenhoven commented:
Thank you! I'll take a look.

Yes I'm still working on this. I got a little behind in keeping it up to date with master. I'm going to rebase it and clean up the history a bit and then it should be good to go.

warmenhoven committed:
libretro core

* Bringing citra libretro implementation over
* libretro: hook up vulkan renderer
* libretro: github actions
* libretro: gyro
* libretro: core options v2
* libretro: on ios turn off shader jit if unavailable
* moltenvk 1.3.0 introduces 8-bit indexes but allocates 16-bit for metal; this ends up allocating stream buffer * 2 = 132MiB. Instead, just use 16-bit indexes. (This will be necessary for standalone when bumping moltenvk version.)

M2GameDeveloper commented:
> I forked the project and did some fixes, there's a working CI build here (scroll to bottom): https://github.com/johndoe6345789/azahar_libretro/actions/runs/20305535751

Here there. I'm on windows. Using retroarch with vulkan. Can't seem to get this to work. Seems to think there is a missing file or something.

Any suggestions on getting this to run? Thank you!

warmenhoven commented on .github/workflows/libretro.yml:
```diff
@@ -0,0 +1,172 @@
+        with:
+          name: ${{ env.OS }}-${{ env.TARGET }}
+          path: ${{ env.BUILD_DIR }}/${{ env.EXTRA_PATH }}/azahar_libretro_android.so
+  linux:
```
I agree that it's overkill to compile all of them every time but think reducing it down to just one introduces too many ways for problems to sneak through. I'd suggest reducing it to iOS and one other, probably Linux.

warmenhoven commented on src/citra_libretro/core_settings.h:
```diff
@@ -0,0 +1,41 @@
+
+struct CoreSettings {
+
+    ::std::string file_path;
```
? Just left over from the citra core.

warmenhoven commented on src/video_core/renderer_vulkan/vk_resource_pool.cpp:
```diff
@@ -159,7 +160,8 @@ void DescriptorHeap::Allocate(std::size_t begin, std::size_t end) {
             break;
         }
-        if (result == vk::Result::eErrorOutOfPoolMemory) {
+        if (result == vk::Result::eErrorOutOfPoolMemory ||
```
The code handles descriptor pool exhaustion by moving to a new pool. eErrorFragmentedPool means the pool has memory but it's too fragmented to satisfy the allocation, which has the same solution as eErrorOutOfPoolMemory. MoltenVK can return eErrorFragmentedPool more frequently than native Vulkan drivers because Metal's memory management differs.

warmenhoven committed:
libretro core: address review feedback

warmenhoven commented on src/audio_core/libretro_sink.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+void LibRetroSink::OnAudioSubmission(std::size_t frames) {
+    std::vector<s16> buffer(frames * 2);
```
`using StereoFrame16 = std::array<std::array<s16, 2>, samples_per_frame>;`

So we need to create a buffer of `(s16 * 2) * frames`. But yes it's not great to then pass size/2 below, I'll fix that.

warmenhoven committed:
libretro: microphone support

warmenhoven commented on src/citra_libretro/emu_window/libretro_window.h:
```diff
@@ -0,0 +1,67 @@
+    int width;
+    int height;
+
+    bool submittedFrame = false;
```
Agreed it should be cleaned up for consistency, but that's a large rename (more than 1000 occurrences across 14 files)....

OpenSauce04 committed:
cmake: Add ENABLE_ROOM_STANDALONE to list of incompatible libretro flags

warmenhoven committed:
libretro: proper initial geometry

MagusRogue commented:
So any update to the playable fork? I'm having issues getting it to run right. it runs really slow and i can't get audio going on it. might just be me though.

OpenSauce04 commented:
> So any update to the playable fork? I'm having issues getting it to run right. it runs really slow and i can't get audio going on it. might just be me though.

Fork?

If you want to test out this feature, please use the actual code branch we're working on.

warmenhoven committed:
libretro: fix software renderer

DavidRGriswold commented on src/citra_libretro/emu_window/libretro_window.cpp:
```diff
@@ -0,0 +1,344 @@
+        break;
+    case Settings::LayoutOption::LargeScreen:
+        if (swapped) { // Bottom screen biggest
+            baseX = Core::kScreenBottomWidth + Core::kScreenTopWidth / 4;
```
Does this mean Large Screen is hardcoded at a 1/4 proportion for now? This should eventually use the Large Screen Proportion setting instead. If that is not available in libretro yet, probably best to still use a variable here and hardcode it with a TODO to fix later...

DavidRGriswold replied:
this also seems to be not taking into account the setting that lets the user put the smaller screen in different positions (specifically on top or on bottom) - I think that's a lower-priority setting to implement for sure than implementing all layout options, but might be worth adding to a list or whatever

PabloMK7 committed:
Merge branch 'master' into warmenhoven/dev/azahar_libretro

warmenhoven commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,125 @@
+    - mac-apple-silicon
+  variables:
+    CORE_ARGS: ${BASE_CORE_ARGS} -DCMAKE_OSX_ARCHITECTURES=x86_64
+    MACOSX_DEPLOYMENT_TARGET: "10.15"
```
No, I believe 10.15 is correct.

warmenhoven commented on src/citra_libretro/citra_libretro.cpp:
```diff
@@ -0,0 +1,715 @@
+#else
+        emu_instance->hw_render.context_type = RETRO_HW_CONTEXT_OPENGL_CORE;
+        emu_instance->hw_render.version_major = 4;
+        emu_instance->hw_render.version_minor = 2;
```
Nope, actually the old citra core incorrectly requested 3.3. Not sure where I got 4.2 from, I'll fix it.

warmenhoven commented on src/common/file_util.cpp:
```diff
@@ -1266,7 +1303,15 @@ std::size_t IOFile::ReadAtImpl(void* data, std::size_t byte_count, std::size_t o
 
     DEBUG_ASSERT(data != nullptr);
 
+#ifdef HAVE_LIBRETRO_VFS
```
Yes you're right, this is definitely not thread safe, good catch, thanks. This was carried over from the old citra core and I did not catch it. I'll add a mutex.

warmenhoven committed:
libretro: address review feedback

OpenSauce04 committed:
.github/libretro.yml: Pin macOS runners at macOS 26

OpenSauce04 committed:
ci: Remove explicit selection of Xcode 16.0

warmenhoven committed:
.github/libretro.yml: remove unnecessary windows builder apt commands

warmenhoven committed:
.github/libretro.yml: bump min macos version to 11.0

OpenSauce04 committed:
ci: Re-enable CI jobs for all libretro cores

This is under the condition that we don't introduce build cache for these builds

OpenSauce04 reviewed (approved):
As far as I can tell, this is good to go. Obviously with a PR this size it's difficult to comprehend the entirety of the changes, and there a couple of things that we need to address in follow-up PRs, but overall this code is in a good state.

@warmenhoven Huge thanks for your help to bring this to fruition! I don't think we could have done it without you.

ACViper commented:
does this work linux? crashes for me
